processmanager/processinfo.go

```diff
@@ -106,128 +106,65 @@ func (pm *ProcessManager) getTSDInfo(pid libpf.PID) *tpbase.TSDInfo {
 	return nil
 }
 
-// updatePidInformation updates pidToProcessInfo with the new information about
-// vaddr, offset, fileID and length for the given pid. If we don't know about the pid yet, it also
-// allocates the embedded map. If the mapping for pid at vaddr with requestedLength and fileID
-// already exists, it returns true. Otherwise false or an error.
+// getPidInformation gets or creates the Pid information for given PID.
 //
 // Caller must hold pm.mu write lock.
-func (pm *ProcessManager) updatePidInformation(pid libpf.PID, m *Mapping) (bool, error) {
-	info, ok := pm.pidToProcessInfo[pid]
-	if !ok {
-		// We don't have information for this pid, so we first need to
-		// allocate the embedded map for this process.
-		var processName string
-		exePath, _ := os.Readlink(fmt.Sprintf("/proc/%d/exe", pid))
-		if name, err := os.ReadFile(fmt.Sprintf("/proc/%d/comm", pid)); err == nil {
-			processName = string(name)
-		}
-
-		envVarMap := make(map[string]string, len(pm.includeEnvVars))
-		if len(pm.includeEnvVars) > 0 {
-			if envVars, err := os.ReadFile(fmt.Sprintf("/proc/%d/environ", pid)); err == nil {
-				// environ has environment variables separated by a null byte (hex: 00)
-				splittedVars := strings.Split(string(envVars), "\000")
-				for _, envVar := range splittedVars {
-					keyValuePair := strings.SplitN(envVar, "=", 2)
-
-					// If the entry could not be split at a '=', ignore it
-					// (last entry of environ might be empty)
-					if len(keyValuePair) != 2 {
-						continue
-					}
-
-					if _, ok := pm.includeEnvVars[keyValuePair[0]]; ok {
-						envVarMap[keyValuePair[0]] = keyValuePair[1]
-					}
-				}
-			}
-		}
-
-		containerID, err := extractContainerID(pid)
-		if err != nil {
-			log.Debugf("Failed extracting containerID for %d: %v", pid, err)
-		}
-
-		info = &processInfo{
-			meta: ProcessMeta{
-				Name:         processName,
-				Executable:   exePath,
-				ContainerID:  containerID,
-				EnvVariables: envVarMap},
-			mappings:         make(map[libpf.Address]*Mapping),
-			mappingsByFileID: make(map[host.FileID]map[libpf.Address]*Mapping),
-			tsdInfo:          nil,
-		}
-		pm.pidToProcessInfo[pid] = info
-
-		// Insert a dummy page into the eBPF map pid_page_to_mapping_info that provides the eBPF
-		// a quick way to check if we know something about this particular process.
-		if err := pm.ebpf.UpdatePidPageMappingInfo(pid, dummyPrefix, 0, 0); err != nil {
-			return false, fmt.Errorf(
-				"failed to update pid_page_to_mapping_info dummy entry for PID %d: %v",
-				pid, err)
-		}
-		pm.pidPageToMappingInfoSize++
-	} else if mf, ok := info.mappings[m.Vaddr]; ok {
-		if *m == *mf {
-			// We try to update our information about a particular mapping we already know about.
-			return true, nil
-		}
+func (pm *ProcessManager) getPidInformation(pid libpf.PID) *processInfo {
```

Comment:
I'm wondering if we could further reduce the required holding of the pm.mu lock. In particular, as reading process metadata from /proc, should not block others from getting the lock in the mean time.